src/v/raft/state_machine_base.h

```diff
@@ -104,4 +115,34 @@ class state_machine_base {
     model::offset _next{0};
 };
 
+/**
+ * This flavor of state machine base allows implementer to opt out from taking
+ * snapshot at arbitrary offset. This way a partition that the STM is based on
+ * will not support the fast partition movement mechanism. When implementing
+ * this state machine make sure to provide an external mechanism that will drive
+ * the snapshot creation.
+ */
+class no_at_offset_snapshot_stm_base : public state_machine_base {
+    /**
+     * Method that will be called whenever a raft snapshot is required
+     */
+    virtual ss::future<iobuf> take_snapshot() = 0;
+
+    snapshot_at_offset_supported supports_snapshot_at_offset() const final {
+        return snapshot_at_offset_supported::no;
+    }
+
+    ss::future<iobuf> take_snapshot(model::offset offset) final {
+        if (offset != last_applied_offset()) {
+            throw std::logic_error(fmt::format(
+              "State machine that do not support taking snapshot at arbitrary "
+              "offset can to take snapshot at requested offset: {}, current "
+              "last applied offset: {}",
+              offset,
+              last_applied_offset()));
+        }
```

Comment:
_apply_units are held by this point, right? Reason I'm asking is because of scheduling points if the units are not held last_applied() may change from L136 and L144. But AFAICT thats not a concern given the units.

Reply:
yes, exactly the units are held in the `state_machine_manager`